kernel/src/fs/utils/flock.rs

```diff
@@ -138,7 +135,17 @@ impl FlockList {
                     return Ok(());
                 }
             }
-            waiter.wait();
+
+            let (waiter, _) = Waiter::new_pair();
+            waiter.pause_until(|| {
+                let list = self.inner.read();
+                if let Some(conflict_lock) = list.iter().find(|l| req_lock.conflict_with(l)) {
+                    conflict_lock.waitqueue.enqueue(waiter.waker());
+                    None
+                } else {
+                    Some(())
+                }
+            })?;
```

Comment:
This seems to be incorrect. If the signal and the wakeup event come at the same time, the wakeup event will be lost because we're returning `EINTR` directly to userspace.

The closure must include the entire `loop` body, not just the conflict checking part.

Reply:
Oh, sorry. For the correctness part, it seems to be fine. As long as the conflict checks are included in the `wait_until` condition, the race conditions should be prevented. Including the action part in the condition is not really necessary to avoid losing the wakeup events.

Perhaps the only downside is that we'll be taking the lock twice after the wakeup event, once for the check and once for the action.

Reply:
Yes, we can include the entire `loop` body to prevent lock twice. We need to change the read lock to write lock in the closure.

Reply:
There are many fast paths in the wait API (https://github.com/asterinas/asterinas/pull/1547#issue-2630289939). It can lead to locking multiple times if the condition is not satisfied. But I think this issue will soon be addressed in that PR.

Reply:
> We need to change the read lock to write lock in the closure.

We use the write lock outside the closure anyway. I guess having the read lock only for the closure does not help much.

> There are many fast paths in the wait API ([#1547 (comment)](https://github.com/asterinas/asterinas/pull/1547#issue-2630289939)). It can lead to locking multiple times if the condition is not satisfied. But I think this issue will soon be addressed in that PR.

 - Yes, #1547 will remove all fast paths in the `waiter` APIs. So fast paths will not be the problem.
 - Instead, the current solution will lead to multiple locks because
   - After entering `pause_until` we have to check the condition, as we generally have to do this to avoid losing wakeup events (this is not a fast path, it is something we have to do for a general API for correctness purposes). However, in this scenario we don't need to, because the condition is checked before entering `pause_until`.
   - Before exiting `pause_until`, we need to check the condition to avoid losing wakeup events. But no action is taken, resulting in another write lock after exiting `pause_until`.
   - So in the current implementation, we'll take the lock at least four times (two for read and two for write) if the resource is not immediately available, but ideally we only need to take the lock two times.

Reply:
However, if we include the entire `loop` in the `cond` closure, we cannot elegantly handle the `is_nonblocking==true` condition, which needs to return `EAGAIN,` but the `cond` closure cannot return it. And there is no API that has an extra `cancel_cond` argument for this purpose.

Reply:
> However, if we include the entire `loop` in the `cond` closure, we cannot elegantly handle the `is_nonblocking==true` condition, which needs to return `EAGAIN,` but the `cond` closure cannot return it. And there is no API that has an extra `cancel_cond` argument for this purpose.

We can do this:
```rust
fn try_set_lock(..., waker: Option<&Waker>) -> Result<()> {
    // do something
    if need_block {
        // enqueue waker
        Err(EAGAIN)
    } else {
        Ok(())
    }
}

fn set_lock(...) -> Result<()> {
    if is_non_blocking {
        return try_set_lock(..., None);
    }

    let (waiter, waker) = Waiter::new_pair();
    waiter.pause_until(|| {
        let result = try_set_lock(..., Some(&waker));
        // ... return `None` if `result` is `Err(EAGAIN)`
    })
}
```

This is a common pattern for handling nonblocking in our current codebase. You can have a look at the use cases of `Pollable::wait_events` for more examples.